### PR TITLE
Add participants to on chain privacy group

### PR DESCRIPTION
**Pull request by iikirilov**

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    if (privacyGroupHeadBlockMap.get(privacyGroupId) != null) {
+      Hash blockHash = privacyGroupHeadBlockMap.get(privacyGroupId);
+      while (blockHash != null) {
+        pmtHashes.addAll(
+            0,
+            privateStateStorage.getPrivateBlockMetadata(blockHash, privacyGroupId).get()
+                .getPrivateTransactionMetadataList().stream()
+                .map(PrivateTransactionMetadata::getPrivacyMarkerTransactionHash)
+                .collect(Collectors.toList()));
+        blockHash = blockchain.getBlockHeader(blockHash).get().getParentHash();
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java:66]: You are working your way from the head down to the first PMT. I think you end up serialising the data starting with the last PMT/PTx. It might be easier for the new member and a bit more logical if the order of the serialised data was from first PMT to the last PMT ...

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java [hunk 2]
```diff
@@ -63,6 +76,29 @@ public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
       return e.getResponse();
     }
 
+    final String addPayloadEnclaveKey;
+    if (privacyController.isGroupAdditionTransaction(privateTransaction)) {
+      final List<Hash> hashes =
+          onChainPrivacyController.buildTransactionList(
```
Comment (pinges) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java]: You have got these three methods in the OnChainPrivacyController, why don't you have another method there that soes all the work and just returns the "addPayloadEnclaveKey"?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetTransactionReceipt.java [hunk 3]
```diff
@@ -106,7 +107,23 @@ public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
...
 
       privateTransaction = PrivateTransaction.readFrom(input);
-      privacyGroupId = receiveResponse.getPrivacyGroupId();
+      if (privateTransaction.getPrivacyGroupId().isPresent()) {
+        PrivacyGroup privacyGroup = null;
+        try {
+          privacyGroup =
+              privacyController.retrievePrivacyGroup(
+                  privateTransaction.getPrivacyGroupId().get().toBase64String());
+        } catch (final EnclaveClientException e) {
```
Comment (pinges) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetTransactionReceipt.java:118]: We should know whether the user wants to use the onchain functionality or not and we should use that information here to make that code clearer.

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/condition/ExpectValidPrivateTransactionReceipt.java [hunk 4]
```diff
@@ -24,21 +24,34 @@
...
 
   @Override
   public void verify(final PrivacyNode node) {
-    assertThat(node.execute(transactions.getPrivateTransactionReceipt(transactionHash)))
-        .isEqualToIgnoringGivenFields(receipt, "commitmentHash");
+    final PrivateTransactionReceipt actualReceipt =
+        node.execute(transactions.getPrivateTransactionReceipt(transactionHash));
+    assertThat(actualReceipt)
+        .usingRecursiveComparison()
+        .ignoringFields("commitmentHash", "logs")
```
Comment (pinges) [inline acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/condition/ExpectValidPrivateTransactionReceipt.java:45]: why can the pmt hash not be compared here?

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/condition/ExpectValidPrivateTransactionReceipt.java [hunk 5]
```diff
@@ -24,21 +24,34 @@
...
+        .isEqualTo(expectedReceipt);
+
+    assertThat(actualReceipt.getLogs().size()).isEqualTo(expectedReceipt.getLogs().size());
+
+    for (int i = 0; i < expectedReceipt.getLogs().size(); i++) {
+      assertThat(actualReceipt.getLogs().get(i))
+          .usingRecursiveComparison()
+          .ignoringFields("blockHash", "blockNumber")
```
Comment (pinges) [inline acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/condition/ExpectValidPrivateTransactionReceipt.java:53]: why ignore the blockHash and the blockNumber?

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/OnChainPrivacyAcceptanceTest.java [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+
+    bob.verify(
+        privateTransactionVerifier.validOnChainPrivacyGroupExists(
+            expectedGroupAfterCharlieIsAdded));
+
+    charlie.verify(
+        privateTransactionVerifier.validOnChainPrivacyGroupExists(
+            expectedGroupAfterCharlieIsAdded));
+
+    final Optional<TransactionReceipt> aliceAddReceipt =
```
Comment (pinges) [inline acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/OnChainPrivacyAcceptanceTest.java:296]: I think to really test that charlie receives all the necessary transactions up to and including the add-transaction and all the transactions after the add we need to "flood" the transaction pool with private transactions before and after the "pause", and before and after the "add". Then we can go through all the receipts and check that all receipts for successful transactions are available to charlie and the same as the ones that alice and bob create.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/privx/PrivxFindPrivacyGroup.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  @Override
+  public String getName() {
+    return RpcMethod.PRIVX_FIND_PRIVACY_GROUP.getMethodName();
+  }
+
+  @Override
+  public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+    LOG.trace("Executing {}", RpcMethod.PRIVX_FIND_PRIVACY_GROUP.getMethodName());
```
Comment (pinges) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/privx/PrivxFindPrivacyGroup.java:53]: With multi-tenancy we have to check whether the current user is allowed to find these privacy groups by checking whether the current user is one of the addresses ... But I think multi-tenancy is not merged in yet?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/OnChainPrivacyPrecompiledContract.java [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+                Bytes.wrap(Base64.getDecoder().decode(addReceiveResponse.getPayload())));
+
+        long lastBlockNumber = -1;
+        for (int i = 0; i < privateTransactionWithMetadataList.size(); i++) {
+          final PrivateTransactionWithMetadata privateTransactionWithMetadata =
+              privateTransactionWithMetadataList.get(i);
+          final PrivateTransactionProcessor.Result result =
+              privateTransactionProcessor.processTransaction(
+                  currentBlockchain,
+                  publicWorldState,
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/OnChainPrivacyPrecompiledContract.java]: The public state to use here is the state at the end of the public Tx just before the PMT that the private Tx was marked by!

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContract.java [hunk 9]
```diff
@@ -129,7 +129,9 @@ public Bytes compute(final Bytes input, final MessageFrame messageFrame) {
     final PrivateTransaction privateTransaction = PrivateTransaction.readFrom(bytesValueRLPInput);
     final WorldUpdater publicWorldState = messageFrame.getWorldState();
     final Bytes32 privacyGroupId =
-        Bytes32.wrap(Bytes.fromBase64String(receiveResponse.getPrivacyGroupId()));
+        privateTransaction.getPrivacyGroupId().isPresent()
+            ? Bytes32.wrap(privateTransaction.getPrivacyGroupId().get())
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContract.java:133]: I cannot see why this change is necessary ...

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  public List<PrivateTransactionMetadata> buildTransactionMetadataList(
+      final Bytes32 privacyGroupId) {
+    final List<PrivateTransactionMetadata> pmtHashes = new ArrayList<>();
+    PrivacyGroupHeadBlockMap privacyGroupHeadBlockMap =
+        privateStateStorage
+            .getPrivacyGroupHeadBlockMap(blockchain.getChainHeadHash())
+            .orElse(PrivacyGroupHeadBlockMap.EMPTY);
+    if (privacyGroupHeadBlockMap.get(privacyGroupId) != null) {
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java:57]: What happens when I'm trying to add someone to a privacy group that does not exist? What should happen? Should we fail?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    this.privacyController = privacyController;
+  }
+
+  public List<PrivateTransactionMetadata> buildTransactionMetadataList(
+      final Bytes32 privacyGroupId) {
+    final List<PrivateTransactionMetadata> pmtHashes = new ArrayList<>();
+    PrivacyGroupHeadBlockMap privacyGroupHeadBlockMap =
+        privateStateStorage
+            .getPrivacyGroupHeadBlockMap(blockchain.getChainHeadHash())
+            .orElse(PrivacyGroupHeadBlockMap.EMPTY);
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java:56]: This means that I'm trying to add someone to a privacy group that does not exist! What is supposed to happen in that case?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+                .getPrivacyGroupHeadBlockMap(blockHash)
+                .orElse(PrivacyGroupHeadBlockMap.EMPTY);
+        if (privacyGroupHeadBlockMap.get(privacyGroupId) != null) {
+          blockHash = privacyGroupHeadBlockMap.get(privacyGroupId);
+        } else {
+          break;
+        }
+      }
+    }
+    return pmtHashes;
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java:78]: variable name? It is not only PMT hashes now!

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java [hunk 13]
```diff
@@ -63,6 +77,31 @@ public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
       return e.getResponse();
     }
 
+    final String addPayloadEnclaveKey;
+    if (privacyController.isGroupAdditionTransaction(privateTransaction)) {
+      final List<PrivateTransactionMetadata> privateTransactionMetadataList =
+          onChainPrivacyController.buildTransactionMetadataList(
+              Bytes32.wrap(privateTransaction.getPrivacyGroupId().get()));
+      if (privateTransactionMetadataList.size() > 0) {
```
Comment (pinges) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java:85]: maybe we should fail if there are no transactions ...

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java [hunk 14]
```diff
@@ -92,8 +127,64 @@ public SendTransactionResponse sendTransaction(
...
-    return enclave.receive(enclaveKey, enclavePublicKey);
+      final String enclaveKey, final Hash blockHash, final String enclavePublicKey) {
+    try {
+      return enclave.receive(enclaveKey, enclavePublicKey);
+    } catch (final EnclaveClientException e) {
+      final PrivacyGroupHeadBlockMap privacyGroupHeadBlockMap =
+          privateStateStorage
+              .getPrivacyGroupHeadBlockMap(blockHash)
+              .orElse(PrivacyGroupHeadBlockMap.EMPTY);
+      final List<Bytes32> addDataKey =
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:138]: should that happen in a PrivacyController? Should that not only happen in the OnChainPrivacyController?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java [hunk 15]
```diff
@@ -92,8 +127,64 @@ public SendTransactionResponse sendTransaction(
...
+          privacyGroupHeadBlockMap.keySet().stream()
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:139]: ... and why is that code different from the code in the OnChainPrivacyControler?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java [hunk 16]
```diff
@@ -113,11 +204,40 @@ public String deletePrivacyGroup(final String privacyGroupId, final String encla
     return enclave.findPrivacyGroup(addresses);
   }
 
+  public List<PrivacyGroup> findOnChainPrivacyGroup(
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:207]: We have to make sure they only get groups back that they are part of. "They" is the enclavePublicKey here. We should check whether "they" are part of the addresses!

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java [hunk 17]
```diff
@@ -164,25 +284,143 @@ public long determineNonce(
 
   private SendResponse sendRequest(
       final PrivateTransaction privateTransaction, final String enclavePublicKey) {
+
+    final Optional<PrivateTransactionProcessor.Result> process =
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:288]: s/process/result/

Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:288]: Could we put the code that is working out the version into a seperate method?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java [hunk 18]
```diff
@@ -164,25 +284,143 @@ public long determineNonce(
...
+            buildCallParams(
+                Bytes.fromBase64String(enclavePublicKey), GET_VERSION_METHOD_SIGNATURE));
+
     final BytesValueRLPOutput rlpOutput = new BytesValueRLPOutput();
+    rlpOutput.startList();
     privateTransaction.writeTo(rlpOutput);
+    process.ifPresent(
+        result -> {
+          if (!result.getOutput().toHexString().equals("0x")) {
+            rlpOutput.writeBytes(result.getOutput());
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:301]: I'm not sure I like to have the "version" just appended to the end of the payload/private transaction. What are the other options that we have? Maybe just append it to the privacy group id? 
Or should we just encode the payload in a nicer way, so that it contains PGID, version, and the private transaction?
We do not have to add the version to the "legacy" private transactions, because the private transaction contains all the recipients, and the 'pantheon' style private transaction don't need it either, because the members cannot change.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/OnChainPrivacyPrecompiledContract.java [hunk 19]
```diff
@@ -223,6 +224,40 @@ public Bytes compute(final Bytes input, final MessageFrame messageFrame) {
...
+
+          final BlockHeader parentBlockHeader =
+              currentBlockchain.getBlockHeader(blockHeader.getParentHash()).orElseThrow();
+
+          // I need to access the world state from the public world state archive
+          final Hash initialStateRoot = parentBlockHeader.getStateRoot();
+
+          // There are the transaction that need to be run in the context of the world state with
+          // state root initialStateRoot
+          final List<Transaction> transactionsToRun =
```
Comment (pinges) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/OnChainPrivacyPrecompiledContract.java]: If there were more than one PMT in the block you would run the same transactions (and more) for the next PMT ...

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/condition/ExpectValidPrivateTransactionReceipt.java (hunk 4 above)
Comment (iikirilov) [reply acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/condition/ExpectValidPrivateTransactionReceipt.java:45]: because the commitment hash will be different every time as the pg id is random

File: acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/condition/ExpectValidPrivateTransactionReceipt.java (hunk 5 above)
Comment (iikirilov) [reply acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/condition/ExpectValidPrivateTransactionReceipt.java:53]: because don't know which block it will be in

File: acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/OnChainPrivacyAcceptanceTest.java (hunk 6 above)
Comment (iikirilov) [reply acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/OnChainPrivacyAcceptanceTest.java:296]: it is hard to do this in the AT

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/privx/PrivxFindPrivacyGroup.java (hunk 7 above)
Comment (iikirilov) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/privx/PrivxFindPrivacyGroup.java:53]: multi-tenancy will introduce complications

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java (hunk 10 above)
Comment (iikirilov) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java:57]: nope - it is fine - will create a new group

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java (hunk 11 above)
Comment (iikirilov) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java:56]: a new group is created

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java (hunk 18 above)
Comment (iikirilov) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:301]: i don't feel that this should be a priority - I believe the impact of refactoring this later will be small
